add switch frames and time

sanglevinh opened the pull request:
- https://github.com/oist/optinist/issues/516

ReiHashimoto commented on frontend/src/components/Workspace/Visualize/Plot/TimeSeriesPlot.tsx:
```diff
@@ -167,7 +196,19 @@ const TimeSeriesPlotImple = React.memo(() => {
+          size: 10,
+          color: 'black',
+        },
+        range: rangeUnit === 'frames' ? [xrange.left, xrange.right] :
+          [xrange.left ? xrange.left / 50 : -1.2 , xrange.right ? xrange.right / 50 : (dataXrange.length / 50 + 3)],
```
declare a constant and leave todo comment for fps for now, to easily replace to retrieved value later.

and use 30 as fps(backend's default fps is 30)

sanglevinh replied:
@ReiHashimoto i fixed it, please check it again

sanglevinh commented on frontend/src/store/slice/VisualizeItem/VisualizeItemUtils.ts:
```diff
@@ -29,7 +29,7 @@ export function isTimeSeriesItem(
 ): item is TimeSeriesItem {
   return (
-    item.itemType === VISUALIZE_ITEM_TYPE_SET.DISPLAY_DATA &&
+    item?.itemType === VISUALIZE_ITEM_TYPE_SET.DISPLAY_DATA &&
```
@ReiHashimoto i fixed it

ReiHashimoto reviewed (changes requested):
timeseries plots have some unexpected peaks.

<img width="489" alt="Screenshot 2023-08-28 at 16 25 33" src="https://github.com/arayabrain/barebone-studio/assets/42664619/2748e6e6-ca99-416a-895b-3105c9fb0abf">

sanglevinh commented:
> timeseries plots have some unexpected peaks.
> 
> <img alt="Screenshot 2023-08-28 at 16 25 33" width="489" src="https://user-images.githubusercontent.com/42664619/263628859-2748e6e6-ca99-416a-895b-3105c9fb0abf.png">

I checked and didn't see it, I think it might be because of the data returned like that
![Screenshot 2023-08-28 at 15 11 28](https://github.com/arayabrain/barebone-studio/assets/109060170/9095f9c4-979a-49ff-ad34-82a377ac88a8)

ReiHashimoto commented:
@sanglevinh 
> I think it might be because of the data returned like that

I think not.

I've run following workflow in develop-main branch.
<img width="893" alt="Screenshot 2023-08-28 at 17 28 38" src="https://github.com/arayabrain/barebone-studio/assets/42664619/1c4e6509-f389-4b59-b218-0175aac7278b">

I got following plots.
<img width="1225" alt="Screenshot 2023-08-28 at 17 23 30" src="https://github.com/arayabrain/barebone-studio/assets/42664619/40ff2c73-4a93-425a-b840-b9df9c18bd1f">

I switched to this PR's branch and reproduced the workflow, then I got following plots.
<img width="1219" alt="Screenshot 2023-08-28 at 17 27 00" src="https://github.com/arayabrain/barebone-studio/assets/42664619/15532132-cf48-4baa-bcff-9eb7ce865491">


Using same json, but show different plots.

sanglevinh commented:
> @sanglevinh
> 
> > I think it might be because of the data returned like that
> 
> I think not.
> 
> I've run following workflow in develop-main branch. <img alt="Screenshot 2023-08-28 at 17 28 38" width="893" src="https://user-images.githubusercontent.com/42664619/263645042-1c4e6509-f389-4b59-b218-0175aac7278b.png">
> 
> I got following plots.
> 
> * left is suite2p_roi's
> * right is caiman_cnmf's
> 
> <img alt="Screenshot 2023-08-28 at 17 23 30" width="1225" src="https://user-images.githubusercontent.com/42664619/263645219-40ff2c73-4a93-425a-b840-b9df9c18bd1f.png">
> I switched to this PR's branch and reproduced the workflow, then I got following plots. <img alt="Screenshot 2023-08-28 at 17 27 00" width="1219" src="https://user-images.githubusercontent.com/42664619/263645826-15532132-cf48-4baa-bcff-9eb7ce865491.png">
> 
> Using same json, but show different plots.

@ReiHashimoto I found the problem and fixed it, please check again

ReiHashimoto commented on frontend/src/components/Workspace/Visualize/Plot/TimeSeriesPlot.tsx:
```diff
@@ -167,7 +198,20 @@ const TimeSeriesPlotImple = React.memo(() => {
+          color: 'black',
+        },
+        range: rangeUnit === 'frames' ? [xrange.left, xrange.right] :
+          [typeof xrange.left !== 'undefined' ? xrange.left / frameRate : -2.5 ,
+            typeof xrange.right !== 'undefined' ? xrange.right / frameRate : (dataXrange.length / frameRate) + 6.8],
```
It looks working finely.

Just to be sure, could you tell me what the values of -2.5 and 6.8 mean?
In other words, the basis for the calculations, etc.

sanglevinh replied:
@ReiHashimoto these two parameters so that when converting from frames to time and vice versa, the graph will not be scaled

ReiHashimoto commented on frontend/src/store/slice/Experiments/ExperimentsSelectors.ts:
```diff
@@ -90,3 +90,11 @@ export const selectExperimentFunctionStatus =
   (uid: string, nodeId: string) => (state: RootState) =>
     selectExperimentFunction(uid, nodeId)(state).hasNWB
+
+export const selectFrameRate = (nodeId: string | null) => (state: RootState) => {
+  if(!nodeId) return 50
+  const list = selectExperimentList(state)
```
specify uid required.
nodeId can be same value among experiments.

e.g.)
create pipeline -> RUN_ALL
then just change frame rate and RUN_ALL
two workflows with different uid, but nodeID is not changed.

sanglevinh replied:
@ReiHashimoto I have fixed it, please check again

ReiHashimoto replied:
@sanglevinh 
Looks fine, thank you.

ReiHashimoto commented on frontend/src/store/slice/Experiments/ExperimentsUtils.ts:
```diff
@@ -11,7 +11,10 @@ export function convertToExperimentListType(
   const experimentList: ExperimentListType = {}
   Object.entries(dto).forEach(([uid, value]) => {
-    experimentList[uid] = convertToExperimentType(value)
+    experimentList[uid] = {
+      ...convertToExperimentType(value),
+      frameRate: convertToExperimentType(value).frameRate
+    }
```
Please refactor here.
Update `ExperimentType` itself.
(`frameRate`'s value should be set in the `convertToExperimentType` method.)

sanglevinh replied:
@ReiHashimoto I fixed it, please help me check
